### PR TITLE
KAS-3608: Blocking users, memberships, and organizations

### PR DESCRIPTION
**Pull request by sergiofenoll**
Related PRs
- [x] https://github.com/kanselarij-vlaanderen/mock-login-service/pull/2
- [x] https://github.com/kanselarij-vlaanderen/mock-login-service/pull/3
- [x] https://github.com/kanselarij-vlaanderen/acmidm-login-service/pull/11
- [x] https://github.com/kanselarij-vlaanderen/app-kaleidos/pull/324

### CODE REVIEW

File: app/styles/addon-overrides/_ember-data-table.scss [hunk 1]
```diff
@@ -1,3 +1,7 @@
 td, th {
   border-radius: 0;
 }
+
+.auk-table > *[data-table], .auk-table .data-table-content {
+  overflow: visible;
+}
```
Comment (sergiofenoll) [inline app/styles/addon-overrides/_ember-data-table.scss:7]: @brenner-company hier de fix om ervoor te zorgen dat de dropdown niet voor een overflow zorgt in een `<DataTable>`.

File: app/controllers/settings/organizations/index.js [hunk 2]
```diff
@@ -1,9 +1,33 @@
...
+    const blocked = await this.store.findRecordByUri('concept', CONSTANTS.USER_ACCESS_STATUSES.BLOCKED);
+    this.organizationBeingBlocked.status = blocked;
+    await this.organizationBeingBlocked.save();
+  }
+
+  @action
+  async unblockOrganization() {
+    const allowed = await this.store.findRecordByUri('concept', CONSTANTS.USER_ACCESS_STATUSES.ALLOWED);
+    this.organizationBeingBlocked.status = allowed;
+    await this.organizationBeingBlocked.save();
```
Comment (erikap) [inline app/controllers/settings/organizations/index.js]: Should we also unblock all memberships of this organization here?  In that case we probably need a loading state in the modal to inform users system is being busy.

File: app/templates/settings/users/user.hbs [hunk 3]
```diff
@@ -75,3 +110,151 @@
...
+        <AuButton
+          @alert={{true}}
+          {{on "click" (queue this.blockMembership (toggle "showBlockMembership" this))}}
+        >
+          {{t "block-membership"}}
+        </AuButton>
+      </Group>
+    </AuToolbar>
+  </:footer>
+</AuModal>
```
Comment (erikap) [inline app/templates/settings/users/user.hbs:149]: Can we create a generic/reusable confirmation modal component with `@onCancel` and `@onConfirm` handlers?

File: app/templates/settings/organizations/index.hbs [hunk 4]
```diff
@@ -19,30 +19,147 @@
...
+              @currentSorting={{this.sort}}
+              @field="identifier"
+              @label={{t "ovo-code"}}
+            />
+            <th>{{t "users-seen"}}</th>
+            <th></th>
           </c.header>
-          <c.body as |organization|>
-            <td>
+          <c.body class="hello" as |organization|>
```
Comment (erikap) [inline app/templates/settings/organizations/index.hbs]: Can be removed? :smiley:

File: app/templates/settings/organizations/index.hbs [hunk 5]
```diff
@@ -19,30 +19,147 @@
...
+            <td class={{if organization.isBlocked "au-u-muted"}}>
```
Comment (erikap) [inline app/templates/settings/organizations/index.hbs:31]: Maybe best to include `status` in the model hook?

File: app/templates/settings/users/index.hbs [hunk 6]
```diff
@@ -44,13 +44,18 @@
           </c.header>
           <c.body as |user|>
             <td data-test-route-settings-users-row-name>
-              <span class="au-u-flex au-u-flex--vertical-center au-u-flex--spaced-small">
+              <span class="au-u-flex au-u-flex--vertical-center au-u-flex--spaced-small {{if user.isBlocked "au-u-muted"}}">
```
Comment (erikap) [inline app/templates/settings/users/index.hbs:47]: Maybe best to include `status` in the model hook?

File: app/templates/settings/users/index.hbs [hunk 7]
```diff
@@ -44,13 +44,18 @@
...
+                  <AuPill @icon="circle-x" @skin="error" @size="small">
+                    {{t "blocked-user"}}
+                  </AuPill>
+                {{/if}}
               </span>
             </td>
             <td class="au-o-flow au-o-flow--tiny">
               {{#each user.memberships as |membership|}}
-                <div class="au-u-flex au-u-flex--vertical-center au-u-flex--spaced-small">
+                <div class="au-u-flex au-u-flex--vertical-center au-u-flex--spaced-small {{if membership.isBlocked "au-u-muted"}}">
```
Comment (erikap) [inline app/templates/settings/users/index.hbs:58]: Maybe best to include `membership.status` in the model hook?

File: app/templates/settings/users/index.hbs [hunk 8]
```diff
@@ -59,6 +64,15 @@
                       {{membership.organization.name}}
                     </p>
                   </div>
+                  {{#if membership.organization.isBlocked}}
+                    <AuPill @icon="circle-x" @skin="error" @size="small">
+                      {{t "blocked-organization"}}
```
Comment (erikap) [inline app/templates/settings/users/index.hbs]: Wasn't the conclusion of the discussion that organization block results in blocking the membership hence no additional pill required here?

File: app/templates/settings/users/index.hbs [hunk 9]
```diff
@@ -90,3 +144,179 @@
...
+          @alert={{true}}
+          @disabled={{is-empty this.membershipsBeingBlocked}}
+          {{on "click" (queue this.blockMemberships this.toggleShowBlockMembership)}}
+        >
+          {{t "block-membership"}}
+        </AuButton>
+      </Group>
+    </AuToolbar>
+  </:footer>
+</AuModal>
```
Comment (erikap) [inline app/templates/settings/users/index.hbs]: For maintainability/readability maybe better to extract this modal to a separate component?  Same for the unblock memberships modal.

**Review by erikap — CHANGES_REQUESTED** [review]

File: app/templates/settings/organizations/index.hbs [hunk 10]
```diff
@@ -19,21 +19,35 @@
...
+            <th>{{t "users-seen"}}</th>
+            <th></th>
           </c.header>
           <c.body as |organization|>
-            <td>
+            <td class={{if organization.isBlocked "au-u-muted"}}>
               <span class="au-u-flex au-u-flex--vertical-center au-u-flex--spaced-small">
                 {{organization.name}}
+                {{#if organization.isBlocked}}
+                <AuPill
```
Comment (erikap) [inline app/templates/settings/organizations/index.hbs]: Indendation not correct

File: app/controllers/settings/organizations/index.js [hunk 11]
```diff
@@ -1,9 +1,78 @@
...
+    // Block all memberships, 10 at a time
+    for (const chonk of chunk(memberships.toArray(), 10)) {
+      yield Promise.all(
+        chonk.map((membership) => {
+          membership.status = blocked;
+          return membership.save();
+        })
+      )
+    }
+  }
```
Comment (erikap) [inline app/controllers/settings/organizations/index.js]: Maybe easier to use a task with [`maxConcurrency` of ember-concurrency](http://ember-concurrency.com/docs/task-concurrency-advanced) that can be reused for blocking as well as unblocking?

```javascript
@task({ maxConcurrency: 10, enqueue: true }) 
*updateMembershipStatus(membership, status) {
  membership.status = status;
  yield membership.save();
}
```

File: app/controllers/settings/organizations/index.js [hunk 12]
```diff
@@ -1,9 +1,81 @@
...
+    this.organizationBeingBlocked.status = allowed;
+    yield this.organizationBeingBlocked.save();
+
+    const memberships = yield this.store.queryAll('membership', {
+      filter: {
+        organization: {
+          ':id:': this.organizationBeingBlocked.id,
+        },
+      },
+    });
```
Comment (erikap) [inline app/controllers/settings/organizations/index.js]: Can we reduce duplication here by adding a `*updateOrganizationStatus(organization, status) { ... }` method?

**Review by erikap — APPROVED** [review]

**Pull request merged**